lib/common/socket.c

```diff
@@ -1573,15 +1573,81 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
 #include "h2o-probes.h"
 #include <sys/stat.h>
 
-static __thread int tracing_map_fd = -1;
-static __thread uint64_t tracing_map_last_attempt = 0;
+static int ebpf_map_create(uint32_t map_type, uint32_t key_size, uint32_t value_size, uint32_t max_entries, const char *map_name)
+{
+    union bpf_attr attr = {
+        .map_type = map_type,
+        .key_size = key_size,
+        .value_size = value_size,
+        .max_entries = max_entries,
+    };
+    strncpy(attr.map_name, map_name, sizeof(attr.map_name));
+    return syscall(__NR_bpf, BPF_MAP_CREATE, &attr, sizeof(attr));
+}
+
+static int ebpf_obj_pin(int bpf_fd, const char *pathname)
+{
+    union bpf_attr attr = {
+        .bpf_fd = (uint32_t)bpf_fd,
+        .pathname = (uint64_t)pathname,
+    };
+    return syscall(__NR_bpf, BPF_OBJ_PIN, &attr, sizeof(attr));
+}
+
+static int ebpf_map_lookup(int fd, const void *key, void *value)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+        .value = (uint64_t)value,
+    };
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+}
+
+static int ebpf_map_delete(int fd, const void *key)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+    };
+    return syscall(__NR_bpf, BPF_MAP_DELETE_ELEM, &attr, sizeof(attr));
+}
+
+int h2o_socket_ebpf_setup(void)
+{
+    if (getuid() != 0) {
+        h2o_error_printf("skipping to set up eBPF maps because bpf(2) requires root privileges\n");
+        return 0;
+    }
 
-static void open_tracing_map(h2o_loop_t *loop)
+    // It creates a pinned BPF object file,
+    // and h2o cannot unlink the file because
+    // h2o drops root privileges after this function
+    // unless its configuration is set to keep the privileges.
+    int fd =
+        ebpf_map_create(BPF_MAP_TYPE_LRU_HASH, sizeof(pid_t), sizeof(uint64_t), H2O_EBPF_RETURN_MAP_SIZE, H2O_EBPF_RETURN_MAP_NAME);
+    if (fd < 0) {
+        h2o_perror("BPF_MAP_CREATE failed");
+        return 0;
+    }
+
+    if (ebpf_obj_pin(fd, H2O_EBPF_RETURN_MAP_PATH) != 0) {
+        h2o_perror("BPF_OBJ_PIN failed");
```

Comment:
We are creating an eBPF map and then calling `ebpf_obj_pin`. What does happen when there is an eBPF map is already registered with the same name? Do we receive an error in that case?

If so, I believe that we should not be reporting an error to the user, as IIUC `get_return_map_fd` function is written to use an existing map.

If not, I believe that we should change the logic to reuse the eBPF map already exists, as the current scheme is designed to allow multiple instances of H2O to use the same return map, and because `get_return_map_fd` is *not* designed to reopen the eBPF map unless it gets removed.

Reply:
I noticed that it just worked even if the pinned map already existed, but in fact `BPF_OBJ_PIN` fails with `EEXIST` in such a case. In that case, however, it seems that the `h2o_socket_ebpf_setup` works as expected, so I've ignored `EEXIST`: 5367b3bace4e52a91682591b24506be4c3e3ed4c

This may be because `BPF_MAP_CREATE` finds an existing map that has the same `attr` first, and only creates a map if it's not found.